L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,387 @@
+----
+* Author(s): Per Grön (pgron@google.com)
+* Approver: a11r
+* Status: Draft
+* Implemented in: C++ (but could be extended to more languages)
+* Last updated: 2017-09-06
+* Discussion at: <google group thread> (filled after thread exists)
+
+
+## Abstract
+
+In an attempt to define an API for gRPC that is both efficient, simple to use and that covers most of gRPC's feature set, this document proposes a new C++ API for gRPC based on (a C++ interpretation of) [Reactive Streams](http://www.reactive-streams.org/).
+
+For more information about Reactive Streams, see for example [a 2 minute introduction to Rx](https://medium.com/@andrestaltz/2-minute-introduction-to-rx-24c8ca793877) or [this one hour video presentation about RxJava](https://www.youtube.com/watch?v=_t06LRX0DV0) (one of the most widely used implementations of Reactive Streams).
+
+
+## Background
+
+gRPC currently offers two versions of its C++ API:
+
+1. An easy to use but relatively slow synchronous version.
+2. A fast but not very user friendly asynchronous version.
+
+The async API is powerful and a good foundation, but it is too low level to be suitable for use as is by application code: It forces application code to worry about manual memory management and designing its own threading model. [This still unfinished PR that attempts to document how to use async bidirectional streaming](https://github.com/grpc/grpc/pull/8934) is an indication that the API simply is difficult to use today.
+
+If a company would adopt C++ gRPC and use it across several teams, the author believes they would have to either build a framework that entirely wraps gRPC's API, or they would end up with code that is very inconsistent between services/teams.
+
+This RFC hopes to make C++ gRPC easier to adopt by offering an API results in applications that are fast and that have simple code at the same time.
+
+
+### Related Proposals: 
+
+* This RFC is closely related to [#7352: Provide a simple event-processing loop for C++ async API](https://github.com/grpc/grpc/issues/7352).
+
+
+## Proposal
+
+This document proposes a new C++ API for gRPC, in addition to the current synchronous and asynchronous APIs. It is based on rather than replaces the current async API. If the C++ version is successful, this proposal could be extended to cover more languages in the future.
+
+The proposal has two main parts: The Reactive Streams library and the new gRPC API.
+
+
+### A C++ Reactive Streams library
+
+[Reactive Streams](http://www.reactive-streams.org/) is currently defined only for Java and Javascript. This project needs a C++ version.
+
+In preparation for this RPC the author has written [a Reactive Streams library in C++ called rs](https://github.com/per-gron/shuriken/tree/master/src/rs). rs is well documented, well tested and implements many of the [ReactiveX operators](http://reactivex.io/documentation/operators.html). Reading rs' documentation may help explain parts of this RFC.
```

Comment:
Would accepting this proposal imply taking a dependency on rs?

Additional forced dependencies on users may be a deal breaker for us.

Reply:
I can see a few ways to do this:

1. Since rs is not used anywhere else and was made basically for this purpose, rs could be made a part of this and included with the rest of the code.

2. If rs is deemed to heavy of a dependency (the implementation + test suite is currently ~7.5kLOC counted with `wc -l`), then we could include only the interface part of rs and leave out all of the operators (they are most of the code). This seems a bit awkward to me but is certainly doable.

3. Given that Publisher is a "concept" (as in forward iterator etc), it is probably technically possible to not have a dependency to rs at all. However, in practice you would have to depend on rs to get things done.